antenna/util.py

```diff
@@ -211,25 +206,51 @@ def _retry_inner(fun):
         @wraps(fun)
         def _retry_fun(*args, **kwargs):
             attempts = 0
+            wait_times = wait_time_generator()
+
             while True:
+                # Grab the next wait time and if there isn't one, then this is
+                # the last attempt
+                try:
+                    next_wait = next(wait_times)
+                except StopIteration:
+                    next_wait = None
+
                 try:
                     ret = fun(*args, **kwargs)
                     if retryable_return is None or not retryable_return(ret):
+                        # This was a successful return--yay!
                         return ret
 
-                    elif max_attempts is not None and attempts >= max_attempts:
-                        raise MaxAttemptsError(
-                            'Maximum retry attempts. Error return.', ret
+                    # The return value is "bad", so we log something and then
+                    # do another iteration.
+                    if module_logger is not None:
+                        module_logger.warning(
+                            '%s: bad return, retry attempt %s',
+                            fun.__qualname__,
+                            attempts
                         )
 
-                except retryable_exceptions:
+                    # If last attempt, then raise MaxAttemptsError
+                    if next_wait is None:
+                        raise MaxAttemptsError('Maximum retry attempts.', ret)
+
+                except retryable_exceptions as exc:
+                    # Retryable exception is thrown, so we log something and
+                    # then do another iteration.
                     if module_logger is not None:
-                        module_logger.exception('retry attempt %s', attempts)
+                        module_logger.warning(
```

Comment:
I ditched `.exception()` because that's pretty spammy. If the last attempt fails, it can print the entire traceback. I switched it to `.warning()` figuring that this shouldn't happen, but it's getting retried, so this particular instance of it isn't an "error" that needs to be dealt with.